### PR TITLE
ENH: Cython interface for special

### PR DESCRIPTION
**Pull request by person142**
This defines typed, scalar versions of many of the functions in
scipy.special and makes them available to cimport in Cython. It uses
fused types so that the functions have the same names and signatures
as their ufunc counterparts.

It piggybacks off of the existing code generation in
generate_ufuncs.py. In particular:
- The process for adding new ufuncs is unchanged
- When a new ufunc is added a Cython version will automatically be
  created (subject to some restrictions mentioned at the end), and the
  documentation and tests will be automatically updated.

This does not attempt to add Cython versions of:
- ~~Ufuncs with multiple return arguments~~ Ufuncs with multiple return arguments are now supported
- Ufuncs that have additional layers of python wrappers.

With more work these restrictions could be removed.

Closes gh-5964.

### CODE REVIEW

**Comment by rgommers** [comment]
This looks quite promising @person142! 

There are 2 test failures for the latest commit, those are real.

**Comment by rgommers** [comment]
Benchmarks are quite consistent; numpy and cython same speed (except ~400ns less overhead for small arrays with cython), Python significantly slower:

    [ 57.14%] ··· Running cython_special.CythonSpecial.time_gamma_d                                      ok
    [ 57.14%] ···· 
                   ========== ====== ========== ========== ==========
                   --                              api               
                   ----------------- --------------------------------
                    argument    N      python     numpy      cython  
                   ========== ====== ========== ========== ==========
                     (100,)     10    10.61μs     2.11μs     1.59μs  
                     (100,)    100    95.18μs    12.64μs    12.19μs  
                     (100,)    1000   890.96μs   122.49μs   117.66μs 
                   ========== ====== ========== ========== ==========

And autogenerating benchmarks is a nice touch.

**Comment by person142** [comment]
Yes, I need to track those failures down. I ran into an issue before that became gh-6208; might be more of the same.

I have noticed one circumstance where cython is slower than numpy, and that's functions with a lot of complex arguments and an underlying c function that's using numpy cdoubles (e.g. the ``airy`` benchmark). I guess the extra time is coming from all the casting back and forth between ``double complex`` and ``cdouble``.

**Comment by pv** [comment]
It would be surprising if the time difference comes from casting, since that should not be very many cpu cycles.

**Comment by person142** [comment]
I looked at the generated cython wrappers for ``pro_rad2`` and ``pro_rad2_cv`` and didn't see anything suspicious, and while gh-6211 doesn't specifically address the test failures it makes me suspect the underlying issue is in ``specfun.f``. (Uninitialized memory getting accessed and making the outputs nondeterministic.)

I marked the radial spheroidal wave functions as known failures; if people are not happy with this I'll dig deeper into the Fortran code and try to pinpoint the error.

**Comment by rgommers** [comment]
@person142 it would be good to advertise this PR on the mailing list. I'm sure some people will be interested.

**Comment by rgommers** [comment]
> I marked the radial spheroidal wave functions as known failures; if people are not happy with this I'll dig deeper into the Fortran code and try to pinpoint the error.

Addressing that separately is ok I think.

File: scipy/special/generate_ufuncs.py [hunk 1]
```diff
@@ -393,13 +393,289 @@ def errprint(inflag=None):
...
+
+CYTHON_SPECIAL_PYX = '''\
+# This file is automatically generated by generate_ufuncs.py.
+# Do not edit manually!
+"""
+================================
+Cython API for Special Functions
+================================
+
+Scalar, typed versions of many of the functions in ``scipy.special``
```
Comment (pv) [inline scipy/special/generate_ufuncs.py]: It could be simpler to just list all the available C signatures, rather than the internal representation used in generate.py --- this way it would be more self-explanatory.

**Comment by person142** [comment]
@pv good point, changed it:

https://circle-artifacts.com/gh/scipy/scipy/174/artifacts/0/home/ubuntu/scipy/doc/build/html-scipyorg/special.cython_special.html

**Comment by ev-br** [comment]
While I won't be able to review the details, I played a bit with this from a user perspective and looks great.  Don't know if this makes this release or the next one, but this definitely should be a highlight of a release.

One longer-term question, not for this PR. Would it be possible to add some magic to dispatch to these functions automatically when a special function is called with scalar arguments?

File: scipy/special/specfun_wrappers.c [hunk 2]
```diff
@@ -77,7 +77,7 @@ npy_cdouble clngamma_wrap( npy_cdouble z) {
 
 npy_cdouble chyp2f1_wrap( double a, double b, double c, npy_cdouble z) {
   npy_cdouble outz;
-  int l1, l0;
+  int l1, l0, isfer;
```
Comment (pv) [inline scipy/special/specfun_wrappers.c]: Initialize `isfer = 0` --- since it's not set by the fortran routine always, the result is undefined.

File: scipy/special/specfun_wrappers.c [hunk 3]
```diff
@@ -88,7 +88,14 @@ npy_cdouble chyp2f1_wrap( double a, double b, double c, npy_cdouble z) {
...
   }
-  F_FUNC(hygfz, HYGFZ)(&a, &b, &c, &z, &outz);
+  F_FUNC(hygfz, HYGFZ)(&a, &b, &c, &z, &outz, &isfer);
+  if (isfer == 3) {
+    sf_error("chyp2f1", SF_ERROR_OVERFLOW, NULL);
+    REAL(outz) = NPY_INFINITY;
+    IMAG(outz) = 0.0;
+  } else if (isfer == 5) {
+    sf_error("chyp2f1", SF_ERROR_LOSS, NULL);
+  }
```
Comment (pv) [inline scipy/special/specfun_wrappers.c]: `else if (isfer != 0) { ... set to nan and emit error ... }`?

File: scipy/special/generate_ufuncs.py [hunk 4]
```diff
@@ -393,13 +395,247 @@ def errprint(inflag=None):
...
+cimport _complexstuff
+cimport scipy.special._ufuncs_cxx
+from . import _ufuncs
+
+ctypedef long double long_double
+ctypedef float complex float_complex
+ctypedef double complex double_complex
+ctypedef long double complex long_double_complex
+
+cdef class _float:
```
Comment (pv) [inline scipy/special/generate_ufuncs.py]: Why are these necessary?

Comment (pv) [inline scipy/special/generate_ufuncs.py]: I think in the tests you will have `type(cython_special._long_double(x).val) is float` iow. this probably doesn't help in choosing the correct signatures.

**Comment by pv** [comment]
Ok, a few comments:

- I think autogeneration for tests and benchmarks is a bit too far --- I'd write these manually.
- The test_cython_special api tests don't seem to pick the function with correct signatures (as noted above, the `_float` et al map to Python floats, so you always check a single specialization). 

I have fixes for the above in my branch, cf https://github.com/person142/scipy/compare/cython-interface...pv:pr-6195

**Comment by pv** [comment]
Apart from that, looks great to me.

**Comment by person142** [comment]
Thanks for the fixes @pv. I moved the known failure check to earlier in the test so that those functions never get evaluated. I've had ``pbwa`` cause a segfault before.

**Comment by pv** [comment]
Other stuff: `cython_special.pxd` doesn't get installed, missing from setup.py

**Comment by person142** [comment]
I think I fixed it.

**Comment by person142** [comment]
And rebased.

**Comment by pv** [comment]
Thanks, merged based on positive feedback.

**Comment by person142** [comment]
Thanks for taking the time to review @pv.
